File: src/services/autoBackupService.ts
/**
 * Automatic daily local backup.
 *
 * All customer data lives on one phone, so the realistic failure mode is not a
 * bug — it is a lost, wiped or broken device with a month of check-ins on it and
 * nobody having pressed "Export Backup". This runs once per calendar day, in the
 * background, and keeps a rolling window of archives on device storage.
 *
 * Rules it deliberately follows:
 *  - never throws: a failed backup must not break app start-up
 *  - never blocks the first paint (the caller schedules it after boot)
 *  - skips when there is nothing to protect, so day one does not create noise
 *  - prunes only its own folder, never the staff's manual exports
 */

import { backupStamp, toDayKey, toIso } from '../utils/date';
import { AppError, logError, toUserMessage } from '../utils/errors';
import { AUTO_BACKUP_DIR, type FileStore } from './fileStore';
import type { BackupService } from './backupService';
import type { CustomerRepository } from '../database/repositories/customerRepository';
import type { SettingsRepository } from '../database/repositories/settingsRepository';

/** How many automatic archives to retain. Roughly a week of history. */
export const AUTO_BACKUP_KEEP = 7;

export const AUTO_BACKUP_ENABLED_KEY = 'auto_backup_enabled';
export const AUTO_BACKUP_DAY_KEY = 'last_auto_backup_day';
/**
 * Timestamp of the last automatic archive.
 *
 * Deliberately separate from `last_backup_at`, which tracks *manual* exports.
 * An automatic archive sits on the same phone as the data it protects, so it
 * must not silence the dashboard's "copy a backup off this device" reminder.
 */
export const AUTO_BACKUP_AT_KEY = 'last_auto_backup_at';

export type AutoBackupStatus =
  | 'disabled'
  | 'up-to-date'
  | 'nothing-to-back-up'
  | 'written'
  | 'failed';

export interface AutoBackupResult {
  status: AutoBackupStatus;
  fileName?: string;
  bytes?: number;
  /** How many old archives were deleted by retention. */
  pruned?: number;
  /** Present when status is 'failed' — safe to show to staff. */
  message?: string;
  /** True when photos had to be left out to keep the archive buildable. */
  dataOnly?: boolean;
}

export class AutoBackupService {
  constructor(
    private readonly backup: BackupService,
    private readonly settings: SettingsRepository,
    private readonly files: FileStore,
    private readonly customers: CustomerRepository,
  ) {}

  /** Missing setting means enabled: existing installs opt in on upgrade. */
  async isEnabled(): Promise<boolean> {
    const stored = await this.settings.get(AUTO_BACKUP_ENABLED_KEY);
    return stored === null || stored === '' ? true : stored === '1';
  }

  async setEnabled(enabled: boolean): Promise<void> {
    await this.settings.set(AUTO_BACKUP_ENABLED_KEY, enabled ? '1' : '0');
  }

  /** `YYYY-MM-DD` of the last successful automatic backup, or ''. */
  async lastRunDay(): Promise<string> {
    return (await this.settings.get(AUTO_BACKUP_DAY_KEY)) ?? '';
  }

  /** ISO timestamp of the last automatic backup, or ''. */
  async lastRunAt(): Promise<string> {
    return (await this.settings.get(AUTO_BACKUP_AT_KEY)) ?? '';
  }

  async listArchives() {
    return this.files.listShared(AUTO_BACKUP_DIR);
  }

  /**
   * Runs a backup if one is due today. Safe to call on every app start.
   */
  async runIfDue(now: Date = new Date()): Promise<AutoBackupResult> {
    try {
      if (!(await this.isEnabled())) return { status: 'disabled' };

      const today = toDayKey(now);
      if ((await this.lastRunDay()) === today) return { status: 'up-to-date' };

      // Nothing to lose yet — don't write an empty archive on a fresh install.
      if ((await this.customers.count()) === 0) return { status: 'nothing-to-back-up' };

      // Years of ID photos can outgrow what the WebView can hold in memory.
      // Losing the records too would be far worse than losing the photos from
      // one archive, so fall back to a data-only backup instead of failing.
      let bytes: Uint8Array;
      let dataOnly = false;
      try {
        bytes = (await this.backup.buildArchive(true)).bytes;
      } catch (e) {
        if (!(e instanceof AppError) || e.code !== 'BACKUP_FAILED') throw e;
        logError('autoBackup.photosTooLarge', e);
        bytes = (await this.backup.buildArchive(false)).bytes;
        dataOnly = true;
      }

      const fileName = `hotel-autobackup-${backupStamp(now)}${dataOnly ? '-data-only' : ''}.zip`;
      await this.files.writeShared(fileName, bytes, AUTO_BACKUP_DIR);

      // Mark the day only after the archive is safely written.
      await this.settings.set(AUTO_BACKUP_DAY_KEY, today);
      await this.settings.set(AUTO_BACKUP_AT_KEY, toIso(now));

      const pruned = await this.prune();
      return { status: 'written', fileName, bytes: bytes.length, pruned, dataOnly };
    } catch (e) {
      // A backup problem must never stop the app from opening.
      logError('autoBackup', e);
      return { status: 'failed', message: toUserMessage(e) };
    }
  }

  /** Deletes the oldest automatic archives beyond the retention window. */
  private async prune(): Promise<number> {
    const archives = await this.files.listShared(AUTO_BACKUP_DIR);
    const excess = archives.slice(AUTO_BACKUP_KEEP);
    for (const file of excess) {
      await this.files.removeShared(file.path);
    }
    return excess.length;
  }
}

File: lib/services/auto_backup_service.dart
import 'dart:typed_data';

import '../database/repositories/customer_repository.dart';
import '../database/repositories/settings_repository.dart';
import '../models/models.dart';
import '../utils/app_date.dart';
import '../utils/app_error.dart';
import 'backup_service.dart';
import 'file_store.dart';

/// Archives kept in the automatic-backup folder — roughly a week's worth.
const autoBackupKeep = 7;

enum AutoBackupStatus { disabled, upToDate, nothingToBackUp, written, failed }

class AutoBackupResult {
  final AutoBackupStatus status;
  final String? fileName;
  final int? bytes;
  final int? pruned;
  final String? message;
  final bool dataOnly;
  const AutoBackupResult({
    required this.status,
    this.fileName,
    this.bytes,
    this.pruned,
    this.message,
    this.dataOnly = false,
  });
}

/// Runs at most once per calendar day, keeping the last [autoBackupKeep]
/// archives in a folder separate from manual exports. Never throws — a
/// failure here must never block app startup.
class AutoBackupService {
  final BackupService _backup;
  final SettingsRepository _settings;
  final FileStore _files;
  final CustomerRepository _customers;
  AutoBackupService(this._backup, this._settings, this._files, this._customers);

  Future<bool> isEnabled() async => ((await _settings.get('auto_backup_enabled')) ?? '1') != '0';

  Future<void> setEnabled(bool enabled) => _settings.set('auto_backup_enabled', enabled ? '1' : '0');

  Future<String?> lastRunDay() => _settings.get('last_auto_backup_day');
  Future<String?> lastRunAt() => _settings.get('last_auto_backup_at');

  Future<List<SharedFile>> listArchives() => _files.listShared(folder: autoBackupDir);

  Future<AutoBackupResult> runIfDue([DateTime? now]) async {
    final clock = now ?? DateTime.now();
    try {
      if (!await isEnabled()) return const AutoBackupResult(status: AutoBackupStatus.disabled);
      if (await lastRunDay() == toDayKey(clock)) {
        return const AutoBackupResult(status: AutoBackupStatus.upToDate);
      }
      final totalCustomers = await _customers.count(const CustomerQuery());
      if (totalCustomers == 0) {
        return const AutoBackupResult(status: AutoBackupStatus.nothingToBackUp);
      }

      var dataOnly = false;
      Uint8List bytes;
      try {
        final built = await _backup.buildArchive(includeImages: true);
        bytes = built.bytes;
      } on AppError catch (e) {
        if (e.code != AppErrorCode.backupFailed) rethrow;
        dataOnly = true;
        final built = await _backup.buildArchive(includeImages: false);
        bytes = built.bytes;
      }

      final fileName = 'hotel-autobackup-${backupStamp(clock)}${dataOnly ? '-data-only' : ''}.zip';
      await _files.writeShared(fileName, bytes, folder: autoBackupDir);

      // Only stamped after the write succeeds, so a failed write doesn't
      // falsely mark the day as already backed up.
      await _settings.set('last_auto_backup_day', toDayKey(clock));
      await _settings.set('last_auto_backup_at', nowIso());

      final pruned = await _prune();
      return AutoBackupResult(
        status: AutoBackupStatus.written,
        fileName: fileName,
        bytes: bytes.length,
        pruned: pruned,
        dataOnly: dataOnly,
      );
    } catch (e) {
      return AutoBackupResult(status: AutoBackupStatus.failed, message: toUserMessage(e));
    }
  }

  /// Prunes only this service's own `auto/` folder — manual exports in the
  /// parent `backups/` folder are never touched.
  Future<int> _prune() async {
    final archives = await listArchives(); // newest first
    if (archives.length <= autoBackupKeep) return 0;
    final excess = archives.sublist(autoBackupKeep);
    for (final file in excess) {
      await _files.removeShared(file.relativePath);
    }
    return excess.length;
  }
}
